### ISSUE TITLE
Loading data with a lot of columns takes a long time at the first call

### ISSUE THREAD

**Issue by bicycle1885** [JuliaData/CSV.jl#388]
I tried to load a ~75MB CSV file with ~20,000 columns and it resulted in a stack overflow as follows:

```
julia> using CSV, DataFrames

shell> gzip -d GSM2230757_human1_umifm_counts.csv.gz

shell> ls -lh GSM2230757_human1_umifm_counts.csv
-rw-r--r--  1 kenta  staff    75M Mar  4 16:42 GSM2230757_human1_umifm_counts.csv

julia> @time CSV.File("GSM2230757_human1_umifm_counts.csv") |> DataFrame;
Internal error: encountered unexpected error in runtime:
StackOverflowError()
process_node! at ./compiler/ssair/ir.jl:875
process_node! at ./compiler/ssair/ir.jl:961 [inlined]
iterate at ./compiler/ssair/ir.jl:1125
iterate at ./compiler/ssair/ir.jl:1133
iterate at ./compiler/ssair/ir.jl:1133
iterate at ./compiler/ssair/ir.jl:1133
iterate at ./compiler/ssair/ir.jl:1133

[ THIS MESSAGE REPEATS THOUSANDS OF TIMES ... ]

iterate at ./compiler/ssair/ir.jl:1133
adce_pass! at ./compiler/ssair/passes.jl:856
run_passes at ./compiler/ssair/driver.jl:126
optimize at ./compiler/optimize.jl:164
typeinf at ./compiler/typeinfer.jl:35
typeinf_edge at ./compiler/typeinfer.jl:497
abstract_call_method at ./compiler/abstractinterpretation.jl:345
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:85
abstract_call at ./compiler/abstractinterpretation.jl:776
abstract_eval_call at ./compiler/abstractinterpretation.jl:805
abstract_eval at ./compiler/abstractinterpretation.jl:890
typeinf_local at ./compiler/abstractinterpretation.jl:1121
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1191
typeinf at ./compiler/typeinfer.jl:14
typeinf_ext at ./compiler/typeinfer.jl:576
typeinf_ext at ./compiler/typeinfer.jl:613
jfptr_typeinf_ext_1.clone_1 at /Applications/Julia-1.1.app/Contents/Resources/julia/lib/julia/sys.dylib (unknown line)
jl_apply_generic at /Users/osx/buildbot/slave/package_osx64/build/src/gf.c:2219 [inlined]
jl_apply at /Users/osx/buildbot/slave/package_osx64/build/src/./julia.h:1571 [inlined]
jl_type_infer at /Users/osx/buildbot/slave/package_osx64/build/src/gf.c:277
jl_compile_method_internal at /Users/osx/buildbot/slave/package_osx64/build/src/gf.c:1819
jl_fptr_trampoline at /Users/osx/buildbot/slave/package_osx64/build/src/gf.c:1863
columns at /Users/kenta/.julia/packages/Tables/qIlOP/src/fallbacks.jl:149 [inlined]
Type at /Users/kenta/.julia/packages/DataFrames/IKMvt/src/other/tables.jl:21
jl_fptr_trampoline at /Users/osx/buildbot/slave/package_osx64/build/src/gf.c:1864
|> at ./operators.jl:813
jl_fptr_trampoline at /Users/osx/buildbot/slave/package_osx64/build/src/gf.c:1864
do_call at /Users/osx/buildbot/slave/package_osx64/build/src/interpreter.c:323
eval_body at /Users/osx/buildbot/slave/package_osx64/build/src/interpreter.c:625
jl_interpret_toplevel_thunk_callback at /Users/osx/buildbot/slave/package_osx64/build/src/interpreter.c:885
Interpreter frame (ip: 4)
Core.CodeInfo(code=Array{Any, (16,)}[
  Core.SlotNumber(id=1) = Expr(:call, Base.gc_num),
  Core.SlotNumber(id=2) = Expr(:call, Base.time_ns),
  Expr(:call, Base.getproperty, :CSV, :(:File)),
  Expr(:call, SSAValue(3), "GSM2230757_human1_umifm_counts.csv"),
  Core.SlotNumber(id=3) = Expr(:call, :|>, SSAValue(4), :DataFrame),
  Expr(:call, Base.time_ns),
  Core.SlotNumber(id=2) = Expr(:call, Base.:(-), SSAValue(6), Core.SlotNumber(id=2)),
  Expr(:call, Base.gc_num),
  Core.SlotNumber(id=4) = Expr(:call, Base.GC_Diff, SSAValue(8), Core.SlotNumber(id=1)),
  Core.SlotNumber(id=2),
  Expr(:call, Base.getproperty, Core.SlotNumber(id=4), :(:allocd)),
  Expr(:call, Base.getproperty, Core.SlotNumber(id=4), :(:total_time)),
  Expr(:call, Base.gc_alloc_count, Core.SlotNumber(id=4)),
  Expr(:call, Base.time_print, SSAValue(10), SSAValue(11), SSAValue(12), SSAValue(13)),
  Expr(:call, Base.println),
  Expr(:return, Core.SlotNumber(id=3))], codelocs=Array{Int32, (16,)}[1, 2, 3, 3, 3, 4, 4, 5, 5, 6, 6, 6, 6, 6, 7, 8], method_for_inference_limit_heuristics=nothing, ssavaluetypes=16, linetjl_interpret_toplevel_thunk at /Users/osx/buildbot/slave/package_osx64/build/src/interpreter.c:894
jl_toplevel_eval_flex at /Users/osx/buildbot/slave/package_osx64/build/src/toplevel.c:764
jl_toplevel_eval at /Users/osx/buildbot/slave/package_osx64/build/src/toplevel.c:773 [inlined]
jl_toplevel_eval_in at /Users/osx/buildbot/slave/package_osx64/build/src/toplevel.c:793
eval at ./boot.jl:328
eval_user_input at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:85
macro expansion at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/REPL/src/REPL.jl:117 [inlined]
#26 at ./task.jl:259
jl_apply at /Users/osx/buildbot/slave/package_osx64/build/src/./julia.h:1571 [inlined]
start_task at /Users/osx/buildbot/slave/package_osx64/build/src/task.c:572
```

The file can be downloaded from <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2230757> (click the download link at the bottom). The next download link may work as well:
<https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSM2230757&format=file&file=GSM2230757%5Fhuman1%5Fumifm%5Fcounts%2Ecsv%2Egz>.

I'm using CSV.jl v0.4.3 and DataFrames.jl v0.17.1.

```
julia> versioninfo()
Julia Version 1.1.0
Commit 80516ca202 (2019-01-21 21:24 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin14.5.0)
  CPU: Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)
Environment:
  JULIA_SHELL = /bin/bash

```

**Comment by nalimilan** [comment]
Probably worth reporting against Julia too, as there's clearly an issue in the compiler.

**Comment by bicycle1885** [comment]
The stack overflow is a problem of Julia and it is already resolved on the development branch. However, I still see significant performance degradation at the first call. Please see the following example; the first call takes more than three minutes while the second call takes only five seconds.
For comparison, [CSVReader.jl](https://github.com/tk3369/CSVReader.jl) takes less than 20 seconds even at the first call.

```
~/w/julia (master|…) $ ./julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.2.0-DEV.419 (2019-03-05)
 _/ |\__'_|_|_|\__'_|  |  Commit 0a471c8c52* (0 days old master)
|__/                   |

julia> using CSV, DataFrames

julia> @time CSV.File("../H5Base/GSM2230757_human1_umifm_counts.csv") |> DataFrame;
186.095883 seconds (80.08 M allocations: 2.850 GiB, 0.91% gc time)

julia> @time CSV.File("../H5Base/GSM2230757_human1_umifm_counts.csv") |> DataFrame;
  4.611685 seconds (28.92 M allocations: 784.372 MiB, 7.41% gc time)

```

**Comment by nalimilan** [comment]
IIRC there are heuristics to avoid compilation when the number of rows is small. Maybe they could be improved by taking into account the ratio between the number of columns and the number of rows? But maybe the compiler could also get faster in that case.

**Comment by quinnj** [comment]
```julia
        # if the # of cells in the file is less than 500K
        columnaccess = (length(names) * length(positions)) < 500_000
```
THe heuristic isn't based on the # of rows, but the total number of "cells" (i.e. rows * columns). The 500K threshold was based on finding the threshold between when dynamic compilation time + parsing time was less than using non-dynamic code path parsing. It seems in this case though, that perhaps the heuristic is failing us.

**Comment by nalimilan** [comment]
Here it looks like we also should look at the ratio between the number of rows and the number of columns.

**Comment by bicycle1885** [comment]
Setting `columnaccess = true` makes loading much faster even at the first call. I think its heuristic should be somehow improved.

```
julia> using CSV, DataFrames
[ Info: Recompiling stale cache file /Users/kenta/.julia/compiled/v1.2/CSV/HHBkp.ji for CSV [336ed68f-0bac-5ca0-87d4-7b16caf5d00b]

julia> @time CSV.File("../H5Base/GSM2230757_human1_umifm_counts.csv") |> DataFrame;
 13.232174 seconds (99.67 M allocations: 2.722 GiB, 9.16% gc time)

```

**Comment by nalimilan** [comment]
@bicycle1885 Do you feel like doing some benchmarking to find out what would be an appropriate threshold for the ratio between the number of columns and the number of rows?

@quinnj Why is the current threshold based on the number of cells? AFAICT the compilation time increases with the number of columns, but the gain due to compilation increases with the number of rows. So I would expect the formula to be `nrows/ncols < X`, not `nrows*ncols < X`.

**Comment by bicycle1885** [comment]
Sorry, but I've decided to create my own package to solve the problem. The package is here: https://github.com/bicycle1885/TableReader.jl. I'll leave this issue open but please feel free to close anytime.

**Comment by quinnj** [comment]
This will be fixed in #395

**Comment by quinnj** [comment]
Fixed in #395
